packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -88,11 +88,160 @@ class AnnotationResult<T> {
+  /// are attached or detached, and detaching a layer from an owner does not
+  /// imply that it has been removed from its parent.
+  LayerHandle? _parentHandle;
+
+  /// Creates a handle that keeps the native resources of this layer from being
+  /// prematurely disposed.
+  ///
+  /// [RenderObject]s automatically store handles to [PictureLayer](s) they draw
+  /// into and dispose of those handles when the [RenderObject.dispose] method
+  /// is called or they otherwise no longer need the reference to that
+  /// [PictureLayer] (such as when they paint a new picture).
+  ///
+  /// A handle is also automatically managed for [RenderObject.layer].
+  ///
+  /// If a [RenderObject] creates layers in addition to its [RenderObject.layer],
+  /// it must create a handle to that layer and dispose of it when the layer is
+  /// no longer needed. For example, if it re-creates or nulls out an existing
+  /// layer in [RenderObject.paint], it should dispose of the handle to the
+  /// old layer. It should also dispose of any layer handles it holds in
+  /// [RenderObject.dispose].
+  LayerHandle createHandle() {
+    assert(!_debugDisposed, 'Attempted to create a handle to a disposed layer.');
+
+    _refCount += 1;
+    return LayerHandle._(this);
+  }
+
+  int _refCount = 0;
+
+  void _unref() {
+    assert(_refCount > 0);
+    _refCount -= 1;
+    if (_refCount == 0) {
+      dispose();
+    }
+  }
+
+  /// Returns the number of objects holding a [LayerHandle] to this layer.
+  ///
+  /// This method throws if asserts are disabled.
+  int get debugHandleCount {
+    late int count;
+    assert(() {
+      count = _refCount;
+      return true;
+    }());
+    return count;
+  }
+
+  /// Clears any retained resources that this layer holds.
+  ///
+  /// This method must dispose resources such as [EngineLayer] and [Picture]
+  /// objects. The layer is still usable after this call, but any graphics
+  /// related resources it holds will need to be recreated.
+  ///
+  /// This method _only_ disposes resources for this layer. For example, if it
+  /// is a [ContainerLayer], it does not dispose resources of any children.
+  /// However, [ContainerLayer]s do remove any children they have when
+  /// this is method is called.
+  ///
+  /// This method automatically gets called when all outstanding [LayerHandle]s
+  /// are disposed. [LayerHandle] objects are typically held by the [parent]
+  /// layer of this layer and any [RenderObject]s that participated in creating
+  /// it.
+  ///
+  /// After calling this method, the object is unusable.
+  @mustCallSuper
+  @protected
+  @visibleForTesting
+  void dispose() {
+    assert(!_debugDisposed);
+    assert(() {
+      assert(
+        _refCount == 0,
+        'Do not directly call dispose on a $runtimeType. Instead, '
+        'use createHandle and LayerHandle.dispose.',
+      );
+      // TODO(dnfield): enable this. https://github.com/flutter/flutter/issues/85066
+      // _debugDisposed = true;
```

Comment:
We could, but that'd mean commenting out more asserts elsewhere instead of this one.

There's only _one_ customer that is actually affected by this in all of customer_testing/google3. I don't plan to leave this commented out for long.